### PR TITLE
ARROW-3487: [Gandiva] simplify fns that return errors

### PR DESCRIPTION
**Pull request by pravindra**

### CODE REVIEW

**Comment by pravindra** [comment]
@praveenbingo @pitrou can you please review this change ?

The travis tests passed except for the R build, which is unrelated to my change.

File: cpp/src/gandiva/native_function.h [hunk 1]
```diff
@@ -39,28 +39,32 @@ enum ResultNullableType {
...
+  static constexpr int32_t NEEDS_FUNCTION_HOLDER = (1 << 2);
+  static constexpr int32_t CAN_RETURN_ERRORS = (1 << 3);
+
   const FunctionSignature& signature() const { return signature_; }
   std::string pc_name() const { return pc_name_; }
   ResultNullableType result_nullable_type() const { return result_nullable_type_; }
-  bool needs_holder() const { return needs_holder_; }
-  bool needs_context() const { return needs_context_; }
+
+  bool NeedsContext() const { return (flags_ & NEEDS_CONTEXT) != 0; }
```
Comment (praveenbingo) [inline cpp/src/gandiva/native_function.h]: nice one :)

**Review by praveenbingo — APPROVED** [review]
Code review +1.

**Comment by pravindra** [comment]
@pitrou - can you please merge this ?

File: cpp/src/gandiva/native_function.h [hunk 2]
```diff
@@ -39,28 +39,32 @@ enum ResultNullableType {
 /// precompiled function.
 class NativeFunction {
  public:
+  // fucntion attributes.
+  static constexpr int32_t NEEDS_CONTEXT = (1 << 1);
+  static constexpr int32_t NEEDS_FUNCTION_HOLDER = (1 << 2);
+  static constexpr int32_t CAN_RETURN_ERRORS = (1 << 3);
```
Comment (wesm) [inline cpp/src/gandiva/native_function.h]: Should use `kMacroName` style

File: cpp/src/gandiva/precompiled/time_test.cc [hunk 3]
```diff
@@ -31,58 +31,22 @@ timestamp StringToTimestamp(const char* buf) {
...
-  const char* date1 = "1972-12-1";
-  cast_to_date = castDATE_utf8(date1, 9, true, (int64_t)&context, &valid);
-  EXPECT_EQ(cast_to_date, 92016000000);
-  EXPECT_EQ(valid, true);
-
-  const char* date2 = "1972222222";
-  cast_to_date = castDATE_utf8(date2, 10, true, (int64_t)&context, &valid);
-  EXPECT_EQ(cast_to_date, 0);
+
+  EXPECT_EQ(castDATE_utf8((int64_t)&context, "1967-12-1", 9), -65836800000);
```
Comment (wesm) [inline cpp/src/gandiva/precompiled/time_test.cc]: C-style casts

File: cpp/src/gandiva/tests/generate_data.h [hunk 4]
```diff
@@ -91,7 +103,7 @@ class FastUtf8DataGenerator : public DataGenerator<std::string> {
     return cur_char_;
   }
 
-  std::random_device random_;
+  Random random_;
   unsigned int max_len_;
   char cur_char_;
 };
```
Comment (wesm) [inline cpp/src/gandiva/tests/generate_data.h]: We should merge all of this data generation with the code in `arrow/test-util.h`, probably create `arrow/util/random.h`. cc @pitrou

**Review by wesm — APPROVED** [review]
+1. left some minor comments

**Comment by pravindra** [comment]
thanks @wesm. I'll address these comments in my next PR.

**Pull request closed without merging**